pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,414 @@
+
+		// If the version annotation doesn't exist, something is wrong with the
+		// Calico Windows node or it needs to be upgraded manually to
+		// a version supported by the calicoWindowsUpgrader.
+		if !exists {
+			return nil, nil, nil, fmt.Errorf("Node %v does not have the version annotation, it might be unhealthy or it might be running an unsupported Calico version.", node.Name)
+		}
+
+		if node.Labels[common.CalicoWindowsUpgradeLabel] == common.CalicoWindowsUpgradeLabelInProgress {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the upgrade in-progress label", node.Name))
+			inProgress[node.Name] = node
+		} else if variant != w.install.Variant || version != expectedVersion {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v doesn't have the latest variant and/or version. variant=%v, expectedVariant=%v, version=%v, expectedVersion=%v", node.Name, variant, w.install.Variant, version, expectedVersion))
+			pending[node.Name] = node
+		} else {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the latest variant and version", node.Name))
+			inSync[node.Name] = node
+		}
+	}
+
+	windowsLog.V(1).Info(fmt.Sprintf("pending=%v, in-progress=%v, in-sync=%v", len(pending), len(inProgress), len(inSync)))
+	return pending, inProgress, inSync, nil
+}
+
+func (w *calicoWindowsUpgrader) upgradeCompleted(node *corev1.Node) bool {
+	_, variant, version := common.GetNodeVariantAndVersion(node)
+	if variant == w.install.Variant && version == w.getExpectedVersion() {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) isUpgradeFromCalicoToEnterprise(node *corev1.Node) bool {
+	_, variant, _ := common.GetNodeVariantAndVersion(node)
+	if variant == operatorv1.Calico && w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) getExpectedVersion() string {
+	if w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return components.EnterpriseRelease
+	}
+	return components.CalicoRelease
+}
+
+func (w *calicoWindowsUpgrader) updateWindowsNodes() {
+	pending, inProgress, inSync, err := w.getNodeUpgradeStatus()
+	if err != nil {
+		windowsLog.Error(err, "Failed to get Windows nodes upgrade status")
+		return
+	}
+
+	// Notify status manager of pending upgrades.
+	for _, node := range pending {
+		w.statusManager.AddWindowsNodeUpgrade(node.Name, false)
+	}
+
+	for _, node := range inProgress {
+		if w.upgradeCompleted(node) {
+			if err := w.finishUpgrade(context.Background(), node); err != nil {
+				// Log the error and continue. We will retry when we update nodes again.
+				windowsLog.Info(fmt.Sprintf("Could not complete upgrade on node %v: %v", node.Name, err))
+				continue
+			}
+			// Successfully finished completing the upgrade. Moving the node
+			// from in-progress to in-sync.
+			inSync[node.Name] = node
+			delete(inProgress, node.Name)
+			w.statusManager.RemoveWindowsNodeUpgrade(node.Name)
```

Comment:
This is a duplicate call - we already call this directly from `finishUpgrade` so don't need to call it again here.